feat: visual explore editing

briangregoryholmes opened the pull request:
https://www.notion.so/rilldata/Visual-Explore-Editor-1a7a88c5e4e84ffab8da33a29987b240?pvs=4

Closes #5806

ericpgreen2 reviewed (changes requested):
Mind adding an e2e test to cover the hot path?

ericpgreen2 reviewed (commented):
Code looks good!

UXQA:
1. When I first hover over the "Preview" button, the dashboard flashes. [See Jam](https://jam.dev/c/360a2b02-0252-45b8-bacb-73611e85b68d)
2. I see "For more options, edit in YAML" even when I'm on the code view. Is that intentional? Clicking on the button doesn't do anything (e.g. focus the code editor).
    <img width="1482" alt="image" src="https://github.com/user-attachments/assets/7b95380e-94dc-4a8b-b6e9-1b9242a24a3a">
3. When I blur the "Display name" input by clicking on the code, the code does not update. Maybe this is a result of contention for the file content, but it'd be nice for the code to update on input blur. It's further confusing when I click outside the code to unfocus it, and the "Display name" still hasn't updated. [See Jam](https://jam.dev/c/6df4eb7d-5b2a-4655-8604-7ecf6b6dc92e)
4. It could merit it's own PR, but for both the Visual Metrics Editor and the Visual Explore Editor, it'd be nice to inject new YAML keys in the same order as the order of the elements in the visual forms. For example, here it'd be nice to see the `display_name` line above the `metrics_view` line.
    <img width="1478" alt="image" src="https://github.com/user-attachments/assets/372712b5-6e81-401a-b3a2-d5dcd77a09cb">

briangregoryholmes commented on web-common/src/features/visual-editing/SidebarWrapper.svelte:
```diff
@@ -0,0 +1,28 @@
+    @apply pt-6 px-5 mb-2;
+  }
+
+  /* h2 {
+    @apply text-sm font-medium;
+  } */
```
Not leftover, just not being used at the moment.

briangregoryholmes commented on web-common/src/features/workspaces/ExploreWorkspace.svelte:
```diff
@@ -14,6 +14,10 @@
   import { runtime } from "@rilldata/web-common/runtime-client/runtime-store";
   import PreviewButton from "../explores/PreviewButton.svelte";
+  import { workspaces } from "@rilldata/web-common/layout/workspace/workspace-stores";
+  import ViewSelector from "../canvas/ViewSelector.svelte";
```
Moved

briangregoryholmes commented on web-common/src/features/workspaces/VisualExploreEditing.svelte:
```diff
@@ -0,0 +1,491 @@
+          window.location.origin + window.location.pathname;
+
+        localStorage.removeItem(`${exploreName}-persistentDashboardStore`);
+        await asyncWait(3000);
```
Removed

ericpgreen2 reviewed (changes requested):
UXQA:
1. The Measures & Dimensions "Expression" option doesn't seem to be working. I see `expr: "*"` is added to the code file, but there's no way to edit the expression in the side panel.
2. There's a family of issues when removing measures/dimensions on specific views:
    i. On the Pivot table, when I remove dimensions from the dashboard via the "Subset" selector, the changes are not immediately reflected. [See Jam](https://jam.dev/c/770eb8f9-e143-4f63-a6b8-5378bc18d0b7). The same problem exists for measures too.

    ii. On the Dimension Detail Table, here's a particular bug where I've opened the Dimension Detail Table, then added/removed dimensions, and the dashboard goes mostly blank: [see Jam](https://jam.dev/c/4d49411a-20d3-444a-9553-14614886a7b8)

    iii. On TDD with a focused dimension, removing the focused dimension from the dashboard triggers an error. [See Jam](https://jam.dev/c/ed807651-9953-4adf-ba5b-323b583de8ce). The same issue exists when removing the focused measure.
3. When I flip "Time zones" or "Time ranges" from Default to Custom, `time_zones` and `time_ranges` entries are added to the code. When I flip back from Custom to Default, the entries remain. When I flip back to Default, I'd expect the `time_zones` and `time_ranges` keys to be cleared (which appears to be the behavior for the Theme config).

briangregoryholmes commented:
1. Fixed
2. Should largely be fixed.
3. This is the intended behavior so that the YAML more explicitly reflects the default options.

briangregoryholmes commented on web-common/src/features/dashboards/workspace/DashboardWithProviders.svelte:
```diff
@@ -0,0 +1,27 @@
+
+{#if metricsViewName}
+  {#key exploreName}
+    <StateManagersProvider {metricsViewName} {exploreName} visualEditing>
```
I would prefer not to perpetuate the provider pattern as my intention is to remove it. It is the source of a lot of unnecessary complexity.

ericpgreen2 replied:
Much agree that the provider pattern isn't great. What makes it worse is that it's duplicated ~5 times through the codebase :), so I was thinking that we could use this helper component as an intermediate step towards removing it.

briangregoryholmes replied:
Will work on that outside of this PR

ericpgreen2 reviewed (changes requested):
I tested removing a measure that was active on the Pivot table, and it felt like I hit a broken state: [Jam](https://jam.dev/c/a01ec62f-856b-446b-a964-8cab20e8e24c)

A few things on the "Expression" input:
1. Page goes blank when I submit an empty Expression. [See Jam](https://jam.dev/c/b3dd5c5c-ba1b-4ffe-bbd9-e60e32654c36).
2. Dashboard appears broken when my expression doesn't match any measures. [See Jam](https://jam.dev/c/dd128e51-4f60-4fba-9f55-9a34f3b88e4d).
3. Using the arrow keys in the Expression input element also pans the time series chart
4. Can we add a link-to-docs to explain the syntax? CC @jkhwu, @ericokuma. I find I am lost as to what is acceptable input here:
    <img width="296" alt="image" src="https://github.com/user-attachments/assets/6d5a4693-c7ec-442a-9042-46dc117e9683">

briangregoryholmes commented:
Could not reproduce the first one, but made a change based off a guess.

1. Fixed
2. Fixed
3. Fixed
4. Would need input. Don't think this should be a blocker for merging, though certainly for release.

ericpgreen2 commented:
Hmm it seems one of those changes regressed the "Subset" UX: [Jam](https://jam.dev/c/d99c038f-d656-408a-a942-df9cf3b54722)

jkhwu commented:
4. Links to docs are now shown in mocks when `expression` is selected for dims and measures:
![Screenshot 2024-11-21 at 9 23 46 AM](https://github.com/user-attachments/assets/32098c15-b304-4145-9935-46f840caae29)
@ericpgreen2 @briangregoryholmes 
These docs about such expressions can hopefully be timed with release @royendo

ericpgreen2 commented on web-common/src/features/dashboards/workspace/Dashboard.svelte:
```diff
@@ -61,6 +61,8 @@
   // Check if the mock user (if selected) has access to the explore
   $: explore = useExploreValidSpec($runtime.instanceId, exploreName);
+
+  $: console.log({ explore: $explore });
```
Leftover log

ericpgreen2 commented on web-common/src/features/visual-editing/MeasureDimensionSelector.svelte:
```diff
@@ -0,0 +1,100 @@
+        onExpressionBlur(expression);
+      }}
+    />
+    <a
+      href="https://docs.rilldata.com/reference/project-files/explore-dashboards"
+    >
```
Nit: open this in a new tab? Because it extends the width of the sidebar, I've found it easy to click by accident.

ericpgreen2 reviewed (approved):
This is great!
